File: 蓝桥杯练习试题/4算法提高/夺宝奇兵/1.java
import java.util.Scanner;
public class Main {
	public static int sum(int[][] bao,int n){
		for (int i = n-2; i >= 0; i--) {
			for (int j = 0; j <= i; j++) {
				int max=bao[i+1][j] > bao[i+1][j+1] ? bao[i+1][j]:bao[i+1][j+1];
				bao[i][j]+=max;
			}
		}
		return bao[0][0];
	}
    public static void main(String[] args) {
		Scanner scan=new Scanner(System.in);
		int n=scan.nextInt();
		int[][] bao=new int[n][n];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j <= i; j++) {
				bao[i][j]=scan.nextInt();
			}
		}
		System.out.println(sum(bao,n));
	}
}

File: 蓝桥杯练习试题/4算法提高/夺宝奇兵/1.c
//夺宝奇兵
#include<stdio.h>
int max(int a,int b)
{
    return a>b?a:b;
}
int main()
{
    int N,i,j;
    int a[101][101]={0};
    int f[101][10]={0};
    
    scanf("%d",&N);
    
    for(i=1;i<=N;i++)
       for(j=1;j<=i;j++)
            scanf("%d",&a[i][j]);
	
	 for(i=N;i>=0;i--)
		for(j=1;j<=i;j++)
		{
			f[i][j]=max( f[i+1][j],f[i+1][j+1] )+a[i][j];
		}
     printf("%d\n",f[1][1]);
    return 0;
}
